app/classifier/tasks/survey/chooser.jsx

```diff
@@ -0,0 +1,290 @@
+import React from 'react';
+import sortIntoColumns from 'sort-into-columns';
+import Translate from 'react-translate-component';
+import TriggeredModalForm from 'modal-form/triggered';
+import counterpart from 'counterpart';
+
+// key codes
+const BACKSPACE = 8;
+const UP = 38;
+const DOWN = 40;
+
+class Chooser extends React.Component {
+  constructor() {
+    super();
+    this.choiceButtons = [];
+    this.handleClearFilters = this.handleClearFilters.bind(this);
+  }
+
+  componentDidMount() {
+    this.sortChoiceButtons();
+  }
+
+  componentDidUpdate() {
+    this.sortChoiceButtons();
+  }
+
+  getFilteredChoices() {
+    return this.props.task.choicesOrder.map((choiceId) => {
+      const choice = this.props.task.choices[choiceId];
+      let rejected = false;
+      Object.keys(this.props.filters).map((characteristicId) => {
+        const valueId = this.props.filters[characteristicId];
+        if (choice.characteristics[characteristicId].indexOf(valueId) > -1) {
+          rejected = true;
+        }
+      });
+      if (rejected) {
+        return null;
+      } else {
+        return choiceId;
+      }
+    }).filter(Boolean);
+  }
+
+  howManyColumns({ length }) {
+    if (length <= 5) {
+      return 1;
+    } else if (length <= 20) {
+      return 2;
+    } else {
+      return 3;
+    }
+  }
+
+  sortChoiceButtons() {
+    // overrides default DOM focus order by sorting the buttons according to task.choicesOrder
+    const newChoiceButtons = [];
+    this.choiceButtons
+      .filter(Boolean)
+      .map((button) => {
+        const choiceID = button.getAttribute('data-choiceID');
+        const index = this.props.task.choicesOrder.indexOf(choiceID);
+        newChoiceButtons[index] = button;
+      });
+    this.choiceButtons = newChoiceButtons.filter(Boolean);
+  }
+
+  whatSizeThumbnails({ length }) {
+    if (length <= 5) {
```

Comment:
This chain of ifs seems insane to me. Could we use `while` perhaps?

Reply:
Isn't `while` used for loops?

Reply:
There's a ton of code in the survey task that made me scratch my head while I was converting it.

Reply:
Indeed, I thought that was a loop. Read too quickly, sorry.
When I've translated components before, I tried to improve the code if possible. You're right though, looks like quite a task in this case.

Reply:
`switch` is what I wanted to say. Would that be a bit neater?

Reply:
I went for a straight rewrite to avoid changing any existing functionality by accident. In the absence of tests, it's hard to tell if a change has broken a function.

Reply:
I appreciate that. I'll review this as is and we should rewrite the survey task in a separate PR.